compiler/passes/src/dead_code_elimination/dead_code_eliminator.rs

```diff
@@ -14,25 +14,89 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use leo_ast::NodeBuilder;
-use leo_span::Symbol;
+use leo_ast::{AccessExpression, Expression, Location, Node, Type};
+use leo_span::{Symbol, sym};
 
 use indexmap::IndexSet;
 
+use crate::{SymbolTable, TypeTable};
+
+#[derive(Debug)]
 pub struct DeadCodeEliminator<'a> {
-    /// A counter to generate unique node IDs.
-    pub(crate) node_builder: &'a NodeBuilder,
     /// The set of used variables in the current function body.
     pub(crate) used_variables: IndexSet<Symbol>,
-    /// Whether or not the variables are necessary.
-    pub(crate) is_necessary: bool,
```

Comment:
I'm for removing the pattern where we set `is_necessary`. It looks like instead we use the traversal to add used variables. Are we sure that there are not defaults in the traversal that might add something unwanted or skip something we'd want?